### PR TITLE
New Post - Getting REST right

### PR DESCRIPTION
**Pull request by jakehall-IW**
Just a little something I've drafted up.

Happy to receive comments, criticism and feedback on the content!

### CODE REVIEW

File: _posts/2017-03-17-getting-rest-right.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+Most agree that you should stick to one or the other across the whole of your API, to keep the amount of learning to a minimum.
+
+Verbs should not show up in your API URLs - if possible. _"But what about `/search`?"_ I hear you cry.
+
+Well if search is specific to an account then maybe having something like `/article?q=search+term` may be more appropriate to limit the results of the article list. Alternatively if you do have global search endpoint, then it's probably appropriate to go for the `/search` endpoint.
+
+## Versioning
+
+Think about versioning your API from the public first release, unless your API is internal only and being consumed only by your own team can you probably get away with no versioning.
```
Comment (flock3) [inline _posts/2017-03-17-getting-rest-right.md]: Get what you're trying to say, perhaps drop the "unless your API is internal only..." or maybe reword to something like: "Unless your API is for internal use only, and will only be used by your team, you should version your API from the first public release"?

**Comment by jakehall-IW** [comment]
Yeah I used `jekyll serve` whilst writing it, but I've done one final check and it looks good.

File: _posts/2017-03-17-getting-rest-right.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+Well if search is specific to an account then maybe having something like `/article?q=search+term` may be more appropriate to limit the results of the article list. Alternatively if you do have global search endpoint, then it's probably appropriate to go for the `/search` endpoint.
+
+## Versioning
+
+Unless your API is for internal use only, and will only be used by your team, you should version your API from the first public release.
+
+There are a couple of different ways you can go about this.
+
+1. In URL versioning (http://api.example.com/v1/article/43)
+   - Probably the most popular, but unfortunately semi-breaks the "URLs represent a resource" principal
```
Comment (gordonbrown) [inline _posts/2017-03-17-getting-rest-right.md]: Not sure if you mean 'principle'?

File: _posts/2017-03-17-getting-rest-right.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    "url": "/v1/article/256"
+  }
+]
+```
+
+The url property of the article object allows the consumer to know how view the detail of the article object, without having to look at your documentation (you did ~~write~~ generate documentation right?). If you're using Postman, you can just click on the link and it will take you straight to the next resource!
+
+## SSL Everywhere
+
+With the invent of Let's Encrypt, there is no longer any excuse not to secure your API endpoint.
```
Comment (gordonbrown) [inline _posts/2017-03-17-getting-rest-right.md]: Not sure if you mean 'invention' or maybe 'advent'?

File: _posts/2017-03-17-getting-rest-right.md (hunk 2 above)
Comment (jakehall-IW) [inline _posts/2017-03-17-getting-rest-right.md]: Well spotted - changed!

File: _posts/2017-03-17-getting-rest-right.md (hunk 3 above)
Comment (jakehall-IW) [inline _posts/2017-03-17-getting-rest-right.md]: I just invent-ed my own word... Changed - thanks!

File: _posts/2017-03-17-getting-rest-right.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+}
+```
+
+## Caching
+
+I'm talking about both types of caching, server side and client side. 
+
+The server should tell the client if they need to get new data. You can use `Last-Modified` or `Etags` headers for this.
+
+Also the server should keep a cache on the data from the datastore, especially if it's immutable or you when the data expires. This will keep response times snappy and consumers happy!
```
Comment (a-h) [inline _posts/2017-03-17-getting-rest-right.md]: `of the data` instead of `on the data`?
`especially if it's immutable or you when the data expires`. I think that there's something missing here.

Finally, the viability of caching is use-case dependent and using it can be more trouble than it's worth, so I'd probably avoid a strong statement that data _should_ be cached.

File: _posts/2017-03-17-getting-rest-right.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+- `GET /user` _List users_
+- `GET /user/<id>` _Get a specific user_
+
+You may have noticed from the above URL schema that I have gone for singular URL nouns across the board. There is still some debate about whether singular or plural nouns are preferable, with plural seeming to win out but with most admitting it's down to individual preference.
+
+Most agree that you should stick to one or the other across the whole of your API, to keep the amount of learning to a minimum.
+
+Verbs should not show up in your API URLs - if possible. _"But what about `/search`?"_ I hear you cry.
+
+Well if search is specific to an account then maybe having something like `/article?q=search+term` may be more appropriate to limit the results of the article list. Alternatively if you do have global search endpoint, then it's probably appropriate to go for the `/search` endpoint.
```
Comment (a-h) [inline _posts/2017-03-17-getting-rest-right.md]: **a** global search endpoint?

File: _posts/2017-03-17-getting-rest-right.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+## SSL Everywhere
+
+With the invention of Let's Encrypt, there is no longer any excuse not to secure your API endpoint.
+
+If you've got HTTPS enabled, and you receive a request to HTTP - just throw an error such as `400 Bad Request {"error": "HTTPS Required"}` rather than redirecting to HTTPS.
+
+## Errors and HTTP status codes
+
+There are more [HTTP status codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes) than you can shake a stick at. Also all good libraries should be able to read these, so you may as well utilise them.
```
Comment (a-h) [inline _posts/2017-03-17-getting-rest-right.md]: Different libraries interpret the codes differently though, e.g. throw an exception on non-200 status codes which can be ... surprising.

File: _posts/2017-03-17-getting-rest-right.md (hunk 4 above)
Comment (JamesBarwell) [inline _posts/2017-03-17-getting-rest-right.md]: If it's immutable, there's a new cache-control header specifically for that now. See: https://hacks.mozilla.org/2017/01/using-immutable-caching-to-speed-up-the-web/

File: _posts/2017-03-17-getting-rest-right.md [hunk 7]
```diff
@@ -0,0 +1,120 @@
+---
+layout: post
+title: Getting REST right
+author: jake_hall
+---
+
+It seems everyone is building a REST API these days - this is great! It's not easy, but there are a few key things you can do to make it easy for your consumers of your API.
```
Comment (JamesBarwell) [inline _posts/2017-03-17-getting-rest-right.md]: Is it worth talking a little more about why I'd want to use REST over say SOAP RPC? Should I just automatically use REST for everything now?

File: _posts/2017-03-17-getting-rest-right.md (hunk 5 above)
Comment (JamesBarwell) [inline _posts/2017-03-17-getting-rest-right.md]: So if we're doing it this way, why wouldn't I just get all of my articles as `GET /article?id=<id>` too? Or `GET /resource?type=article&id=<id>`? Of course, this search recommendation is the standard way of doing things, so I'm not saying it's wrong, but it's not clear why this different (RPC?) style is being used. So it goes back to my earlier question about why I'm using REST in the first place, is it always appropriate for everything, what are the pros and cons vs RPC, etc.

File: _posts/2017-03-17-getting-rest-right.md [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+Unless your API is for internal use only, and will only be used by your team, you should version your API from the first public release.
+
+There are a couple of different ways you can go about this.
+
+1. In URL versioning (http://api.example.com/v1/article/43)
+   - Probably the most popular, but unfortunately semi-breaks the "URLs represent a resource" principle
+2. Custom header (X-API-Version: v1)
+   - Developers have to learn a new header just to use your API.
+3. Accept header (Accept: application/vnd.article-v1+json)
+   - Developers can't explore your API just using a browser.
```
Comment (willm) [inline _posts/2017-03-17-getting-rest-right.md]: Not necessarily true, most browsers will send */*  in the `Accept` header as a fallback, a good default might be to return the latest version in this case.

File: _posts/2017-03-17-getting-rest-right.md (hunk 4 above)
Comment (sjiveson) [inline _posts/2017-03-17-getting-rest-right.md]: I've got my pedant hat on, but it really should be *TLS Everywhere* seeing as no-one should actually be using SSL anymore. The problem is, everyone still thinks of it as SSL. Also, links for Postman and Let's Encrypt might be useful?

Comment (jakehall-IW) [inline _posts/2017-03-17-getting-rest-right.md]: Yep absolutely, my bad! 

Good shout on the links - added!

File: _posts/2017-03-17-getting-rest-right.md (hunk 8 above)
Comment (jakehall-IW) [inline _posts/2017-03-17-getting-rest-right.md]: Absolutely - I've added a comment about that

File: _posts/2017-03-17-getting-rest-right.md (hunk 7 above)
Comment (jakehall-IW) [inline _posts/2017-03-17-getting-rest-right.md]: Think it's a good one for another post? Trying to keep the length down and focussing mainly on having chosen to go with REST, how to do it as well as it can be done!

There are benefits to both, it depends what kind of data you're providing.

File: _posts/2017-03-17-getting-rest-right.md (hunk 5 above)
Comment (jakehall-IW) [inline _posts/2017-03-17-getting-rest-right.md]: SP fixed.

Agreed that the search example looks more like RPC than REST, but I don't know of a better way of doing it in the REST world (happy to learn different though!). Worth a discussion in a REST vs RPC article though?

File: _posts/2017-03-17-getting-rest-right.md (hunk 4 above)
Comment (jakehall-IW) [inline _posts/2017-03-17-getting-rest-right.md]: Fixed spelling and missing words, also didn't know about this new header - cheers (and I've added it in)!

**Comment by flock3** [comment]
@jakehall-IW I'm happy to merge if you're happy?

**Comment by jakehall-IW** [comment]
Yep, ready for merge - thanks all!

**Comment by flock3** [comment]
Merged 🥇 👍
